### PR TITLE
Add warning for horizon timeout

### PR DESCRIPTION
**Pull request by vwallfahrer**
Horizon timeout is impacted by keystone token timeout since when the
token gets outdated, then the user is automatically unable to use
horizon. This patch includes a check for the timeout of horizon and
keystone barclamps.

### CODE REVIEW

File: crowbar_framework/app/models/horizon.rb [hunk 1]
```diff
@@ -75,6 +75,15 @@ def create_proposal
   def validate_proposal_after_save proposal
     validate_one_for_role proposal, "horizon-server"
 
+    horizon_timeout =  proposal["attributes"]["nova_dashboard"]["session_timeout"]
+    keystone_proposal = Proposal.where(barclamp: "keystone", name: "default").first
+    keystone_timeout = keystone_proposal["attributes"]["keystone"]["token_expiration"]
+
+    # keystone_timeout is in seconds and horizon_timeout is in minutes
+    if horizon_timeout * 60 > keystone_timeout
+        validation_error("Horizon timeout is bigger than keystone token expiration and these may cause problems.")
```
Comment (vuntz) [inline crowbar_framework/app/models/horizon.rb]: I'd suggest rephrasing to: "The session timeout cannot be larger than the Keystone token expiration timeout."

**Comment by rsalevsky** [comment]
Hrm... code looks good so far but it's the same code twice. Maybe but it in an validation helper method and just call the method.

**Comment by abel-navarro** [comment]
@rsalevsky you are right there is some code repeated (3 lines). Since the code is repeated in two different files and it's very specific for the validation, I'd say it would make it harder to read if you need to go to a third file for just this three lines. Would the code had been in the same file or had been more critical, I would totally go for a non-repetition approach.

**Comment by abel-navarro** [comment]
@vuntz 's comment seems to have disappeared after the pull request update. This is his original comment:
> I'd suggest rephrasing to: "The session timeout cannot be larger than the Keystone token expiration timeout."

Btw, I agree with the rephrasing. Although I'd explicitly mention this is horizon's session timeout:
"The **horizon** session timeout cannot be larger than the Keystone token expiration timeout."

**Comment by abel-navarro** [comment]
Just thought of another corner case to try: what happens if we have keystone barclamp installed but not the horizon one? Probably there should be a check when retrieving the proposal to see it's not empty.

**Comment by abel-navarro** [comment]
I checked our default values for the expirations in cloud5:
* horizon: 24 hours (1440 minutes)
* keystone: 4 hours (14400 seconds)

So, the situation that triggers this problem is there by default. There won't be any warning in crowbar because it will only show upon a change. We will need to change the default values.

If we change the default values, should we also change any documentation for the end user?

**Comment by abel-navarro** [comment]
I still haven't fully backported this to cloud5 and probably cannot finish this soon. FTR, this is what the code would look like for cloud5:
```ruby
    horizon_timeout = proposal["attributes"]["nova_dashboard"]["session_timeout"]
    keystone_proposal = ProposalObject.find_proposal("keystone", "default")
    unless keystone_proposal.nil?
      keystone_timeout = keystone_proposal["attributes"]["keystone"]["token_expiration"]

      # keystone_timeout is in seconds and horizon_timeout is in minutes
      if horizon_timeout * 60 > keystone_timeout
        validation_error("The horizon session timeout cannot be larger than the Keystone token expiration timeout.")
      end
    end
```

**Comment by vuntz** [comment]
> If we change the default values, should we also change any documentation for the end user?

I don't think we document that, so go ahead.

**Comment by abel-navarro** [comment]
Regarding the default timeouts, what about making both (horizon + keystone) to be 4 hours?

**Comment by vuntz** [comment]
> Regarding the default timeouts, what about making both (horizon + keystone) to be 4 hours?

We need to make the settings consistent, at least. I'd possibly go for 24 hours, but no strong opinion.

File: crowbar_framework/app/models/keystone_service.rb [hunk 2]
```diff
@@ -83,6 +83,15 @@ def validate_proposal_after_save proposal
       validation_error("API version 3 or newer is required when enabling domain specific drivers.")
     end
 
+    keystone_timeout = proposal["attributes"]["keystone"]["token_expiration"]
+    horizon_proposal = Proposal.where(barclamp: "nova_dashboard", name: "default").first
+    horizon_timeout = horizon_proposal["attributes"]["nova_dashboard"]["session_timeout"]
```
Comment (vuntz) [inline crowbar_framework/app/models/keystone_service.rb]: In this case, there's no guarantee that we have a horizon proposal, so we should check if `horizon_proposal` is `nil`.

File: crowbar_framework/app/models/keystone_service.rb [hunk 3]
```diff
@@ -83,6 +83,15 @@ def validate_proposal_after_save proposal
...
+
+    # keystone_timeout is in seconds and horizon_timeout is in minutes
+    if horizon_timeout * 60 > keystone_timeout
+      validation_error("Horizon timeout is bigger than keystone token expiration and these may cause problems.")
```
Comment (vuntz) [inline crowbar_framework/app/models/keystone_service.rb]: s/these/this/

Comment (vuntz) [inline crowbar_framework/app/models/keystone_service.rb]: Also, s/keystone/Keystone/

File: crowbar_framework/app/models/horizon.rb [hunk 4]
```diff
@@ -75,6 +75,15 @@ def create_proposal
   def validate_proposal_after_save proposal
     validate_one_for_role proposal, "horizon-server"
 
+    horizon_timeout = proposal["attributes"]["nova_dashboard"]["session_timeout"]
+    keystone_proposal = Proposal.where(barclamp: "keystone", name: "default").first
+    keystone_timeout = keystone_proposal["attributes"]["keystone"]["token_expiration"]
+
+    # keystone_timeout is in seconds and horizon_timeout is in minutes
+    if horizon_timeout * 60 > keystone_timeout
+      validation_error("Horizon timeout is bigger than keystone token expiration and these may cause problems.")
```
Comment (vuntz) [inline crowbar_framework/app/models/horizon.rb]: (same comments for the string)

File: crowbar_framework/app/models/keystone_service.rb (hunk 3 above)
Comment (vuntz) [inline crowbar_framework/app/models/keystone_service.rb]: And finally, s/timeout/session timeout/. Sorry for all the comments, should have made just one :-)

**Comment by vuntz** [comment]
Also needs a rebase

File: crowbar_framework/app/models/keystone_service.rb (hunk 3 above)
Comment (dirkmueller) [inline crowbar_framework/app/models/keystone_service.rb]: you can edit comments ;)

File: crowbar_framework/app/models/horizon.rb (hunk 4 above)
Comment (dirkmueller) [inline crowbar_framework/app/models/horizon.rb]: I think this should be more specific. it doesn't cause "problems" per se, its just that a Horizon timeout bigger than the keystone token timeout is ignored. 

So I'd rephrase it similar to 

"Setting the Horizon timeout (%d minutes) longer than the Keystone token expiration timeout (%d minutes) is not supported. Please raise the Keystone token timeout in the keystone proposal."

Comment (abel-navarro) [inline crowbar_framework/app/models/horizon.rb]: I'd remove the last part of the message ("please raise..."). The user may want to lower the horizon timeout instead.

File: crowbar_framework/app/models/keystone_service.rb [hunk 5]
```diff
@@ -83,6 +83,21 @@ def validate_proposal_after_save proposal
...
     end
 
+    keystone_timeout = proposal["attributes"]["keystone"]["token_expiration"]
+    horizon_proposal = Proposal.where(barclamp: "nova_dashboard", name: "default").first
+    unless horizon_proposal.nil?
+      horizon_timeout = horizon_proposal["attributes"]["nova_dashboard"]["session_timeout"]
+
+    timeout_warning = <<-EOF
+Setting the Horizon timeout (#{horizon_timeout} minutes) longer than the Keystone token expiration timeout (#{keystone_timeout * 60} minutes) is not supported.
+Please lower the Horizon token timeout.
```
Comment (vuntz) [inline crowbar_framework/app/models/keystone_service.rb]: Here I would change the error and say "Please increase the Keystone token expiration timeout." (since the user is editing the keystone proposal)

File: crowbar_framework/app/models/keystone_service.rb [hunk 6]
```diff
@@ -83,6 +83,21 @@ def validate_proposal_after_save proposal
       validation_error("API version 3 or newer is required when enabling domain specific drivers.")
     end
 
+    keystone_timeout = proposal["attributes"]["keystone"]["token_expiration"]
+    horizon_proposal = Proposal.where(barclamp: "nova_dashboard", name: "default").first
+    unless horizon_proposal.nil?
+      horizon_timeout = horizon_proposal["attributes"]["nova_dashboard"]["session_timeout"]
+
+    timeout_warning = <<-EOF
```
Comment (vuntz) [inline crowbar_framework/app/models/keystone_service.rb]: Indeed, missing two spaces here :-)

File: crowbar_framework/app/models/horizon_service.rb [hunk 7]
```diff
@@ -83,6 +83,19 @@ def validate_proposal_after_save proposal
       if keystone["attributes"][ks_svc.bc_name]["api"]["version"].to_f < 3.0
         validation_error("Multi domain support requires enabling Keystone V3 API in the keystone proposal first.")
       end
+
+    horizon_timeout = proposal["attributes"]["nova_dashboard"]["session_timeout"]
+    keystone_proposal = Proposal.where(barclamp: "keystone", name: "default").first
+    keystone_timeout = keystone_proposal["attributes"]["keystone"]["token_expiration"]
+
+    timeout_warning = <<-EOF
+Setting the Horizon timeout (#{horizon_timeout} minutes) longer than the Keystone token expiration timeout (#{keystone_timeout * 60} minutes) is not supported.
```
Comment (vuntz) [inline crowbar_framework/app/models/horizon_service.rb]: Hrm, shouldn't that be `keystone_timeout / 60`?

**Comment by tboerger** [comment]
I would prefer the messages to be part of the i18n file as they are getting touched anyway.

**Comment by abel-navarro** [comment]
since @vwallfahrer is sick, I'm continuing the PR in this new one: https://github.com/crowbar/crowbar-openstack/pull/51

**Comment by tboerger** [comment]
Continued in #51
